### ISSUE TITLE
Licence GPLV3 incomptible with secure device

### ISSUE THREAD

**Issue by fulup-bzh** [rpm-software-management/microdnf#79]
I would be interested to contribute with my team to 'microdnf' unfortunately GPLV3 is incompatible with secureboot and devices that require a strong security model. It is easy to understand that a car/train/ship is not like a phone and require a strict enforcement on packages that can/cannot be install. This enforcement is unfortunately incompatible with GPLV3.

We already published DNF plugins for Redpack the container model with use for embedded systems  https://github.com/redpesk/red-pak . Nevertheless we like the idea of removing Python from embedded targets and would be more than happy to contribute to your project if we could use it inside car, boat,... 

Question:  micro-yuminst is GPLV2 which is OK to secure devices. Unfortunatly  MicroDNF switched to GPLV3. Would you conciser to release microdnf under GPLV2 ? If not we will have to restart from strach or from micro-yum.

**Comment by dmach** [comment]
We have a mid-term plan to reimplement microdnf from scratch as a CLI interface to the new libdnf:
https://github.com/rpm-software-management/libdnf/tree/dnf-5-devel
The new libdnf is licensed under LGPLv2+ and the binaries under GPLv2+.
I suppose we could have a new working microdnf by the end of the year.
Would that work for you?

**Comment by fulup-bzh** [comment]
Daniel,
    
    
    Thank you for this information.
    
    
    The GPLV2 licence and your target date
      fine for us. For our embedded container project we had to patch
      current libdnf. I will ask my team to check this new version of
      libdnf. Hopefully we wont have to hack this one, and potentially
      we will propose a contribution to support a hierary of
      master/slave dnf caches.
    
    
    Fulup
    
    
    
    On 11/06/2020 12:53, Daniel Mach wrote:
    
    
      
      We have a mid-term plan to reimplement microdnf from scratch as
        a CLI interface to the new libdnf:
        https://github.com/rpm-software-management/libdnf/tree/dnf-5-devel
        The new libdnf is licensed under LGPLv2+ and the binaries under
        GPLv2+.
        I suppose we could have a new working microdnf by the end of the
        year.
        Would that work for you?
      —
        You are receiving this because you authored the thread.
        Reply to this email directly, view it on GitHub, or unsubscribe.
      [
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/rpm-software-management/microdnf/issues/79#issuecomment-642568611",
"url": "https://github.com/rpm-software-management/microdnf/issues/79#issuecomment-642568611",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by dmach** [comment]
@fulup-bzh you mentioned "we had to patch current libdnf".
Are any of these changes upstreamable?
Can you eventually write down your use cases and create preferably bugs in Red Hat Bugzilla or issues on github?

**Comment by fulup-bzh** [comment]
Daniel,
    
    
    I hope we can upstream all or work. 
      Nevertheless because of the new libdnf, I'm planning to place all
      our effort directly on V5 version. 
    
    All the work we are doing for embedded
      light weight container is opensource (apache-V2) and will do
      everything possible to use an unmodified version of libdnf as we
      already use an unmodified version of librpm.
    
    
    Fulup
    
    
    
    On 11/06/2020 20:01, Daniel Mach wrote:
    
    
      
      @fulup-bzh
        you mentioned "we had to patch current libdnf".
        Are any of these changes upstreamable?
        Can you eventually write down your use cases and create
        preferably bugs in Red Hat Bugzilla or issues on github?
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view it on GitHub, or unsubscribe.
      [
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/rpm-software-management/microdnf/issues/79#issuecomment-642843155",
"url": "https://github.com/rpm-software-management/microdnf/issues/79#issuecomment-642843155",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by dmach** [comment]
I'm fine with changing the license to GPLv2+.
@Conan-Kudo you seem to be the only external contributor in microdnf, others are Red Hatters (or were at the time of their contributions). Could you confirm that you're fine with the license change?

```
$ git shortlog -s -n -e
    42  Jaroslav Rohel <jrohel@redhat.com>
    15  Igor Gnatenko <ignatenko@redhat.com>
     9  Jaroslav Mracek <jmracek@redhat.com>
     7  Aleš Matěj <amatej@redhat.com>
     3  Lukáš Hrázký <lhrazky@redhat.com>
     3  Pavla Kratochvilova <pkratoch@redhat.com>
     2  Daniel Mach <dmach@redhat.com>
     1  Colin Walters <walters@verbum.org>
     1  Daniel Mach <daniel.mach@gmail.com>
     1  Jonathan Dowland <jdowland@redhat.com>
     1  Lukáš Hrázký <lukkash@email.cz>
     1  Marek Blaha <mblaha@redhat.com>
     1  Neal Gompa <ngompa13@gmail.com>
     1  Nicola Sella <nsella@redhat.com>
```

**Comment by Conan-Kudo** [comment]
I'm absolutely fine with the license change.

**Comment by fulup-bzh** [comment]
Thank you all to accept moving the licence, this will allow us to restart our effort on MicroDNF.

**Comment by wi24rd** [comment]
Interesting, microdnf is just a package manager as we all know. It looks like [a mere aggregation in your container system](https://opensource.com/article/18/1/containers-gpl-and-copyleft),
I think, it just requre you give user freedom right to install their own modified microdnf by using such like your modified microdnf, which is the only diffrence between 2 and 3?

**Comment by fulup-bzh** [comment]
@wi24rd, On most critical infrastructure project, the list of packages you may install is restricted to what ever has been certified/authorized. As a result GPLV3 is not allowed, this is especially true for a module that itself install software and require a higher level of trust than simple applications. It is easy to understand why a train, a submarine, a car, ... require a stick control enforcement of every component a user may install. This is the exact reason why  reasons Linus and the Linux kernel stuck with GPLv2 instead of GPLv2+ https://linuxreviews.org/Linux_Remains_Licensed_Under_The_GNU_GPL_v2_And_New_Contributions_Remain_Subject_To_It

**Comment by Conan-Kudo** [comment]
This issue is now fixed with the release of Micro DNF version 3.5.1.

(As an aside, the industry needs to get a move on and start figuring out how to accept GPLv3+ software. I disagree with the underlying premise that GPLv3's terms are antithetical to critical infrastructure solutions, and indeed many folks in the industry have concurred with the assessment that having software that can't be modified on-device is flawed.)
